src/main/network/dns.rs

```diff
@@ -0,0 +1,170 @@
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::Write;
+use std::net::Ipv4Addr;
+use std::os::fd::AsRawFd;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use rustix::fs::MemfdFlags;
+use shadow_shim_helper_rs::HostId;
+
+#[derive(Debug)]
+struct Database {
+    name_index: HashMap<String, Arc<Record>>,
+    addr_index: HashMap<Ipv4Addr, Arc<Record>>,
+}
+
+#[derive(Debug)]
+struct Record {
+    id: HostId,
+    addr: Ipv4Addr,
+    name: String,
+}
+
+#[derive(Debug)]
+pub struct DnsBuilder {
+    db: Database,
+}
+
+impl DnsBuilder {
+    pub fn new() -> Self {
+        Self {
+            db: Database {
+                name_index: HashMap::new(),
+                addr_index: HashMap::new(),
+            },
+        }
+    }
+
+    pub fn register(&mut self, id: HostId, addr: Ipv4Addr, name: String) {
+        if !addr.is_loopback() && !addr.is_unspecified() {
+            let record = Arc::new(Record {
+                id,
+                addr,
+                name: name.clone(),
+            });
+            self.db.name_index.insert(name, record.clone());
+            self.db.addr_index.insert(addr, record);
```

Comment:
Should this method return an error if there was already an entry with the same name or address or host ID?

Reply:
Looks good to me!